### DISCUSSION TITLE
Solution without port forwarding?

### DISCUSSION THREAD

**Discussion by md27182** [coder/code-server#2864] [Q&A] [answered]
I don't have much experience with networking, but I've been following the setup guide the best I can for my use case. 

Instead of using a Google Cloud server, I would like to run code-server on a desktop machine at my office. I would like to be able to access it from anywhere using a browser.

I'm not sure what the rules are for port forwarding at my office and the security risks are a bit intimidating, so I wanted to see if there was a way to set this up without using port forwarding. Will the Let's Encrypt + Caddy/NGINX method in the guide achieve this or will that only work for Google Cloud instances?

It seems like what I want to do is definitely [possible](https://medium.com/swlh/code-anywhere-your-own-vs-code-server-2397b195cb15) with NGROK, but the latency is making me look for another option.

I also found this blog [post](https://jerrington.me/posts/2019-01-29-self-hosted-ngrok.html) which looks promising, but I'm way out of my depth at this point and would appreciate some guidance!

**Comment by md27182** [comment, marked as answer]
Update: I just set up NGROK and it actually works just fine, the latency isn't an issue!

**Reply by jsjoeio** [reply]
Wohoo! Glad to hear that worked! That's awesome :D

**Reply by md27182** [reply]
It's incredible! Huge shoutout to the teams that worked on code-server and ngrok!

**Reply by bpmct** [reply]
Glad to hear you got this working with ngrok!

Wanted to point out that we also have a built in proxy if you run code-server with the `--link` flag! Once you're logged in, you can also access any other ports running on your machine with /proxy/[port]. We have more info on this in the [FAQ](https://github.com/cdr/code-server/blob/main/docs/FAQ.md#how-do-i-securely-access-web-services).

**Reply by md27182** [reply]
This sounds interesting, but the FAQ section you linked is going a bit over my head. Are you saying that I wouldn't need to use ngrok? Does this solution require an ssh connection between my machines or for me to own a domain?

**Reply by bpmct** [reply]
So if you start code-server like so, it will launch code-server with a unique for you, like ngrok. It will also ask you to log in with GitHub so that is secure. No need to maintain a SSH connection

```
code-server --link
```

**Reply by md27182** [reply]
When I run code-server --link, this is the output I get:

```
[2021-03-17T01:00:36.225Z] info  Using user-data-dir ~/.local/share/code-server
[2021-03-17T01:00:36.234Z] info  Using config file ~/.config/code-server/config.yaml
[2021-03-17T01:00:36.234Z] info  HTTP server listening on http://127.0.0.1:XXXXX (randomized by --link)
[2021-03-17T01:00:36.234Z] info    - Authentication is disabled (disabled by --link)
[2021-03-17T01:00:36.234Z] info    - Not serving HTTPS (disabled by --link)
[2021-03-17T01:00:36.752Z] info  Detected an acceptable latency of 62ms
[2021-03-17T01:02:16.846Z] info  Failed to login: unable to login: read session token: read msg: failed to read JSON message: failed to get reader: failed to read frame header: EOF
[2021-03-17T01:02:16.849Z] error coder cloud agent exited with 1
```

**Reply by bpmct** [reply]
Looks like something went wrong here... Try running

```
code-server --link testurl
```

**Reply by md27182** [reply]
That worked beautifully! So cool that it uses GitHub to log in.

With ngrok, the connection was going through their servers. Whose servers is the connection going through now?

Update: now ```code-server --link``` is working just on its own.

**Reply by bpmct** [reply]
The connection goes through our servers, which are geographically distributed. So strange about that error. @sreya may be able to help debug this if it happens again.

**Reply by md27182** [reply]
Sounds good thanks for your help! One last question - is this feature going to eventually become the $35/month paid service on coder.com, or will it remain free?

**Reply by bpmct** [reply]
This will remain free. The $35/mo plan relates to [Coder](https://coder.com/docs), our dev platform for teams.

The source code for the --link server is open source, so if we ever decide to stop supporting it, another instance could be hosted: https://github.com/cdr/cloud-agent

**Reply by md27182** [reply]
Sweet! Okay last, last question - is there a way to get this --link flag added to the default systemd service that was created when I installed code-server?

**Reply by bpmct** [reply]
Yep, it can be done. The easiest way is to edit your code-server config and add to `~/.config/code-server.config.yaml`:

```
link: true
```
You could also edit the systemd service definition and add the --link flag

If it is a new machine, you will have to use journalctl or something to get the systemd logs for the OAuth URL the first time. I also created [deploy-code-server](https://github.com/cdr/deploy-code-server) that adds code-server with link to any Ubuntu VM and forwards port 80 to the --link URL, so there's no need to SSH in or view logs.
